compiler/src/dotty/tools/dotc/transform/ArrayConstructors.scala

```diff
@@ -31,17 +33,34 @@ class ArrayConstructors extends MiniPhase {
     if (tree.fun.symbol eq defn.ArrayConstructor) {
       val TypeApply(tycon, targ :: Nil) = tree.fun
       expand(targ.tpe, tree.args)
-    } else if ((tree.fun.symbol.maybeOwner eq defn.ArrayModule) && (tree.fun.symbol.name eq nme.ofDim) && !tree.tpe.isInstanceOf[MethodicType]) {
-      val Apply(Apply(TypeApply(_, List(tp)), _), _) = tree
-      val cs = tp.tpe.widen.classSymbol
-      tree.fun match {
-        case Apply(TypeApply(t: Ident, targ), dims)
-          if !TypeErasure.isGeneric(targ.head.tpe) && !ValueClasses.isDerivedValueClass(cs) =>
-          expand(targ.head.tpe, dims)
-        case Apply(TypeApply(t: Select, targ), dims)
-          if !TypeErasure.isGeneric(targ.head.tpe) && !ValueClasses.isDerivedValueClass(cs) =>
-          Block(t.qualifier :: Nil, expand(targ.head.tpe, dims))
-        case _ => tree
+    } else if (tree.fun.symbol.maybeOwner eq defn.ArrayModule) {
+      if ((tree.fun.symbol.name eq nme.ofDim) && !tree.tpe.isInstanceOf[MethodicType]) {
+        val Apply(Apply(TypeApply(_, List(tp)), _), _) = tree
+        val cs = tp.tpe.widen.classSymbol
+        tree.fun match {
+          case Apply(TypeApply(t: Ident, targ), dims)
+            if !TypeErasure.isGeneric(targ.head.tpe) && !ValueClasses.isDerivedValueClass(cs) =>
+            expand(targ.head.tpe, dims)
+          case Apply(TypeApply(t: Select, targ), dims)
+            if !TypeErasure.isGeneric(targ.head.tpe) && !ValueClasses.isDerivedValueClass(cs) =>
+            Block(t.qualifier :: Nil, expand(targ.head.tpe, dims))
+          case _ => tree
+        }
+      } else {
+        // Optimize `Array.apply([....])` and `Array.apply(..., [....])` into `[...]`
+        tree match {
+          case Apply(Apply(appMeth, Apply(wrapRefArrayMeth, (seqLit: tpd.JavaSeqLiteral) :: Nil) :: Nil), (ct: Apply) :: Nil)
+              if appMeth.symbol.name == nme.apply && appMeth.symbol.owner == defn.ArrayModule &&
+                 defn.WrapArrayMethods().contains(wrapRefArrayMeth.symbol) &&
+                 ct.symbol.maybeOwner.companionModule == defn.ClassTagModule =>
```

Comment:
Add a comment explaining why this is the correct thing to check.

Reply:
Modified the code and added comments